### PR TITLE
Using a more generic Gtk detection method.

### PR DESCRIPTION
**Pull request by zliuva**
Using Gtk::Settings that reads the `Net/ThemeName` key from XSettings.

https://www.freedesktop.org/wiki/Specifications/XSettingsRegistry/

XSettings are respected by most Gtk desktops and this allows us to
support a wider range of desktops.

Theoretically any desktop with an XSettings daemon should be supported.

Tested:
* i3 + xfsettingsd (XSettings daemon from Xfce 4)
* i3 + gsd-xsettings (Gnome Settings Daemon)

### CODE REVIEW

**Comment by zliuva** [comment]
Hi there!

I understand there is a pending PR #50 using the FreeDesktop dark mode preference.

However, as a LTS user on stable distros, I do not see myself able to use this on my work computer any time soon.

However, I did discover a more generic way that should work with many GTK desktops that implements the XSettings protocol.

Thought this'd be useful for other folks stuck in the same situation.

File: plugin/src/test/kotlin/com/github/weisj/darkmode/platform/linux/gnome/GnomeNativeTest.kt [hunk 1]
```diff
@@ -76,7 +78,7 @@ class GnomeNativeTest {
         }!!
 
         "gsettings set $settingsPath $settingsKey Adwaita".runCommand()
-        assertEquals("Adwaita", service.currentGtkTheme)
```
Comment (zliuva) [inline plugin/src/test/kotlin/com/github/weisj/darkmode/platform/linux/gnome/GnomeNativeTest.kt:79]: I removed this line because the existing test case does not quite make sense to me:

we decrement the `countDownLatch` in the eventhandler's callback.

however, with this line here, we don't really wait for the latch to assert, making the test itself subject to race conditions.

File: base/src/main/java/com/github/weisj/darkmode/platform/LibraryUtil.java [hunk 2]
```diff
@@ -38,5 +38,6 @@ private LibraryUtil() {
     public static final boolean isM1 = SystemInfo.isMac && CpuArch.isArm64();
     public static final boolean isMacOSMojave = SystemInfo.isMacOSMojave;
     public static final boolean isMacOSCatalina = SystemInfo.isMacOSCatalina;
-    public static final boolean isGnome = SystemInfo.isGNOME;
+    // really any GTK would do (`isGnome` named for historical reasons)
+    public static final boolean isGnome = SystemInfo.isGNOME || SystemInfo.isXfce || SystemInfo.isI3;
```
Comment (weisJ) [inline base/src/main/java/com/github/weisj/darkmode/platform/LibraryUtil.java:42]: No need to keep the old name. Please rename to `isGTK`

File: plugin/src/test/kotlin/com/github/weisj/darkmode/platform/linux/gnome/GnomeNativeTest.kt [hunk 3]
```diff
@@ -43,9 +43,10 @@ class GnomeNativeTest {
 
     private fun currentGtkTheme(): String {
         val theme = "gsettings get $settingsPath $settingsKey".runCommand()
+        //val theme = "xfconf-query -c xsettings -p /Net/ThemeName".runCommand()
```
Comment (weisJ) [inline plugin/src/test/kotlin/com/github/weisj/darkmode/platform/linux/gnome/GnomeNativeTest.kt:46]: Why is this commented out?

File: plugin/src/test/kotlin/com/github/weisj/darkmode/platform/linux/gnome/GnomeNativeTest.kt [hunk 4]
```diff
@@ -62,6 +63,7 @@ class GnomeNativeTest {
         GnomeLibrary.get()
 
         "gsettings set $settingsPath $settingsKey Adwaita-dark".runCommand()
+        //"xfconf-query -c xsettings -p /Net/ThemeName -s Adwaita-dark".runCommand()
```
Comment (weisJ) [inline plugin/src/test/kotlin/com/github/weisj/darkmode/platform/linux/gnome/GnomeNativeTest.kt:66]: Same here

**Review by weisJ — COMMENTED** [review]
I think it would be good to rename all `Gnome**` classes/files to read `GTK**`. Same goes for the subproject.

File: plugin/src/test/kotlin/com/github/weisj/darkmode/platform/linux/gnome/GnomeNativeTest.kt (hunk 3 above)
Comment (zliuva) [reply plugin/src/test/kotlin/com/github/weisj/darkmode/platform/linux/gnome/GnomeNativeTest.kt:46]: on my computer I use `xfconf-query` from XFCE but I don't see the need to change the default test based on gnome.

I left it as a comment to show that alternative utilities would also work.

**Comment by zliuva** [comment]
> I think it would be good to rename all Gnome** classes/files to read GTK**

I'm ok with that. Hopefully the diff is still obvious of the actual functional change.

**Comment by weisJ** [comment]
> > I think it would be good to rename all Gnome** classes/files to read GTK**
> 
> I'm ok with that. Hopefully the diff is still obvious of the actual functional change.

As long as you keep the renaming of files in a separate commit it should be fine.

**Comment by weisJ** [comment]
Looks like the CI is failing because the build actions need to be updated

**Comment by zliuva** [comment]
> Looks like the CI is failing because the build actions need to be updated

lol I used ripgrep to find `Gnome` references but I just learned it ignores dot directories by default!

File: plugin/build.gradle.kts [hunk 5]
```diff
@@ -21,6 +21,10 @@ intellij {
 tasks.withType<PatchPluginXmlTask> {
     changeNotes.set(
         """
+        v1.7.0
+        <ul>
+            <li>Generic GTK desktops support. (XSettings based)</li>
+        </ul>
```
Comment (weisJ) [inline plugin/build.gradle.kts:27]: Please make this its own commit

File: gradle.properties [hunk 6]
```diff
@@ -11,7 +11,7 @@ systemProp.org.gradle.internal.publish.checksums.insecure = true
 org.gradle.jvmargs                                        = -noverify
 
 # Version
-auto-dark-mode.version                                    = 1.6.2-2022.1
+auto-dark-mode.version                                    = 1.7.0-2022.1
```
Comment (weisJ) [inline gradle.properties:14]: This should also go into the release note commit.

File: linux/gtk/src/main/java/com/github/weisj/darkmode/platform/linux/gtk/GtkSettings.kt [hunk 7]
```diff
@@ -31,18 +31,18 @@ import com.github.weisj.darkmode.platform.settings.*
...
 @AutoService(SettingsContainerProvider::class)
-class GnomeSettingsProvider : SingletonSettingsContainerProvider({ GnomeSettings }, enabled = LibraryUtil.isGnome)
+class GtkSettingsProvider : SingletonSettingsContainerProvider({ GtkSettings }, enabled = LibraryUtil.isGtk)
 
 data class GtkTheme(val name: String) : Comparable<GtkTheme> {
     override fun compareTo(other: GtkTheme): Int = name.compareTo(other.name)
 }
 
-object GnomeSettings : DefaultSettingsContainer(identifier = "gnome_settings") {
+object GtkSettings : DefaultSettingsContainer(identifier = "gtk_settings") {
```
Comment (weisJ) [inline linux/gtk/src/main/java/com/github/weisj/darkmode/platform/linux/gtk/GtkSettings.kt:40]: This will invalidate existing stored settings. Either keep the identifier the old one with a comment explaining why or advance the settings version in https://github.com/weisJ/auto-dark-mode/blob/master/plugin/src/main/java/com/github/weisj/darkmode/AutoDarkModeOptions.kt

File: plugin/src/test/kotlin/com/github/weisj/darkmode/platform/linux/gnome/GnomeNativeTest.kt (hunk 3 above)
Comment (weisJ) [reply plugin/src/test/kotlin/com/github/weisj/darkmode/platform/linux/gnome/GnomeNativeTest.kt]: As these comments are removed in the folowing commit please drop them from this one :)

File: README.md [hunk 8]
```diff
@@ -8,7 +8,7 @@ operating system settings. The plugin distinguishes between `Light`, `Dark` and
 the theme used for each mode can be customized.
 This plugin currently works for Windows and macOS.
 
-Linux support is both limited and experimental. At the moment, the only Linux desktop environment supported by this plugin is Gnome.
+Linux support is both limited and experimental. At the moment, Linux desktop environments based on Gtk and has an [XSettings](https://www.freedesktop.org/wiki/Specifications/xsettings-spec/) daemon are supported.
```
Comment (weisJ) [inline README.md:11]: This sounds odd. How about "Linux desktop environments based on Gtk, which have an [XSettings](https://www.freedesktop.org/wiki/Specifications/xsettings-spec/) daemon are supported."

**Review by weisJ — COMMENTED** [review]
Some last comment. Otherwise this looks good to me. Very good work. Thank for tackling this!

File: linux/gtk/src/main/java/com/github/weisj/darkmode/platform/linux/gtk/GtkSettings.kt (hunk 7 above)
Comment (zliuva) [reply linux/gtk/src/main/java/com/github/weisj/darkmode/platform/linux/gtk/GtkSettings.kt:40]: ah i was not aware of this side effect. I think keeping the name with a comment is the least disruptive solution.

File: README.md (hunk 8 above)
Comment (zliuva) [reply README.md:11]: I wonder if I should expand on that:

1. An XSettings daemon (e.g. `gsd-xsettings` for Gnome; `xfsettingsd` for Xfce; or even standalone ones like https://github.com/derat/xsettingsd) is required because it is the one that actually sends out the notifications when a theme (amont other settings) changes;
2. Conversely, you can run a desktop environment without XSettings daemon (e.g. to my knowledge LXDE does not have one by default), in which case theme changes are not "real time" (i.e. they usually  takes effect only for new apps launched, since running apps do not get notified)
3. I haven't verified it, but I think even non-Gtk native environments like KDE _should_ work with this, _if_ it runs such a daemon (e.g. https://github.com/KDE/xsettings-kde)

so really the dominating factor is not Gtk (we just use it to get notified), but rather the XSettings daemon (who actually sends out the notifications)

Comment (weisJ) [reply README.md:11]: Thank you for the clarification. Nonetheless the setence as it is isn't grammatically correct. I also think it would be worth while to elaborate a bit on what is needed to use the plugin. This could also be added to https://github.com/weisJ/auto-dark-mode/blob/master/plugin/src/main/resources/META-INF/plugin.xml

**Comment by zliuva** [comment]
hi @weisJ thanks for your comments. I've rebased the PR and:
* reverted the settings' `identifier`
* squashed implementations into a single commit
* left version bumps into its own commit

while implementing this change I also learned quite a bit about desktop environments (vs window managers) and the XSettings daemon. the main take-away I have is that as long as there is such a daemon running and it's signaling the changes according to the spec, this new method of detecting current theme (and change of themes) should work.

I've updated README.md to summarize such learnings in the most technically accurate way possible, while keeping it short (enough).

**Comment by weisJ** [comment]
Thanks a lot for tackling this 👍🏼 

> while implementing this change I also learned quite a bit about desktop environments (vs window managers) and the XSettings daemon. the main take-away I have is that as long as there is such a daemon running and it's signaling the changes according to the spec, this new method of detecting current theme (and change of themes) should work.

In this case it could be possible to allow users to enable monitoring on their own risk (i.e. they make sure that a compatible daemon is installed). But this is something for another time :)

**Comment by zliuva** [comment]
Thank you for your feedback :) 

> Thanks a lot for tackling this 👍🏼
> 
> > while implementing this change I also learned quite a bit about desktop environments (vs window managers) and the XSettings daemon. the main take-away I have is that as long as there is such a daemon running and it's signaling the changes according to the spec, this new method of detecting current theme (and change of themes) should work.
> 
> In this case it could be possible to allow users to enable monitoring on their own risk (i.e. they make sure that a compatible daemon is installed). But this is something for another time :)

Yeah that's what I'm thinking as a potential future improvement.
